client/network/src/protocol/notifications/behaviour.rs

```diff
@@ -916,9 +991,99 @@ impl Notifications {
 
 			// Any state other than `Incoming` is invalid.
 			peer => {
-				error!(target: "sub-libp2p",
-					"State mismatch in libp2p: Expected alive incoming. Got {:?}.",
-					peer);
+				error!(
+					target: LOG_TARGET,
+					"State mismatch in libp2p: Expected alive incoming. Got {:?}.", peer
+				);
+				debug_assert!(false);
+			},
+		}
+	}
+
+	/// Protocol rejected inbound substream.
+	fn protocol_report_reject(&mut self, index: sc_peerset::IncomingIndex) {
+		self.peerset_report_reject(index);
+	}
```

Comment:
Likely `protocol_report_reject` should differ from `peerset_report_reject` in a way that we must notify `ProtocolController` that the peer has dropped. Otherwise it's going to think that we are connected to that peer.